File: orchestrator/app/runtime/context_builder.py
"""Assembles what the runtime knows before any AI runs.

Context management is an AI concern, so the policy lives here, not in the
gateway. The gateway ships a generous message window + the total count;
this module decides how much of it each subscription tier actually uses.
One dict to tune quality vs. cost — never touch the frontend for it.

This module knows nothing about prompts. It only gathers information.
"""

from dataclasses import dataclass, field

from app.models import ChatMessage, ChatStreamRequest, Subscription
from app.tools.retrieval import RetrievedContext


@dataclass(frozen=True)
class ContextPolicy:
    recent_messages: int
    use_summary: bool


CONTEXT_POLICY: dict[Subscription, ContextPolicy] = {
    "free": ContextPolicy(recent_messages=6, use_summary=True),
    "pro": ContextPolicy(recent_messages=20, use_summary=True),
}

# Refresh the rolling summary on the first exchange, then every 5 exchanges
SUMMARY_REFRESH_EVERY = 10


@dataclass
class RuntimeContext:
    request: ChatStreamRequest
    policy: ContextPolicy

    query: str
    history: list[ChatMessage]  # trimmed, excludes the new user message
    summary: str | None
    should_update_summary: bool

    # Filled in by later runtime stages.
    retrieved: RetrievedContext | None = None
    program_templates: str | None = None
    extra: dict = field(default_factory=dict)


def build_context(request: ChatStreamRequest) -> RuntimeContext:
    policy = CONTEXT_POLICY[request.user_context.subscription]

    history = request.messages[:-1]
    if policy.recent_messages:
        history = history[-policy.recent_messages :]

    # Count as it will be after this turn persists (user + assistant).
    count_after_turn = request.total_message_count + 2
    should_update_summary = (
        count_after_turn == 2 or count_after_turn % SUMMARY_REFRESH_EVERY == 0
    )

    return RuntimeContext(
        request=request,
        policy=policy,
        query=request.query,
        history=history,
        summary=request.summary if policy.use_summary else None,
        should_update_summary=should_update_summary,
    )

File: java-orchestrator/src/main/java/com/powerlifting/orchestrator/chat/runtime/ContextBuilder.java
package com.powerlifting.orchestrator.chat.runtime;

import com.powerlifting.orchestrator.chat.model.ChatMessage;
import com.powerlifting.orchestrator.chat.model.ChatStreamRequest;
import java.util.List;
import org.springframework.stereotype.Component;

/**
 * Assembles the {@link RuntimeContext} for a turn: which prior messages to keep,
 * whether to use the stored summary, and whether this turn should refresh it.
 *
 * <p>Context sizing is a subscription-tier decision (see {@link ContextPolicy}),
 * so it lives here rather than in the gateway — the gateway sends a generous
 * window and this trims it to what the tier pays for.
 */
@Component
public class ContextBuilder {

    public RuntimeContext build(ChatStreamRequest request) {
        ContextPolicy policy = ContextPolicy.forSubscription(request.userContext().subscription());

        // Everything except the new user message, then trimmed to the window.
        List<ChatMessage> history = request.messages().subList(0, request.messages().size() - 1);
        if (policy.recentMessages() > 0 && history.size() > policy.recentMessages()) {
            history = history.subList(history.size() - policy.recentMessages(), history.size());
        }

        // Counted as it will be once this turn persists (user + assistant).
        int countAfterTurn = request.totalMessageCount() + 2;
        boolean shouldUpdateSummary =
                countAfterTurn == 2 || countAfterTurn % ContextPolicy.SUMMARY_REFRESH_EVERY == 0;

        return new RuntimeContext(
                request,
                policy,
                request.query(),
                history,
                policy.useSummary() ? request.summary() : null,
                shouldUpdateSummary);
    }
}
